File: leetcode-04/solutions/cpp/0792.cpp
class Solution {
 public:
  int numMatchingSubseq(string S, vector<string>& words) {
    int ans = 0;

    vector<vector<int>> map(26);
    for (int i = 0; i < S.length(); i++) map[S[i] - 'a'].push_back(i);

    for (string& word : words) ans += isMatch(word, map);

    return ans;
  }

 private:
  unordered_map<string, bool> m;

  bool isMatch(string& word, vector<vector<int>>& map) {
    if (m.count(word)) return m[word];

    int prev = -1;

    for (char c : word) {
      vector<int>& p = map[c - 'a'];
      auto it = lower_bound(p.begin(), p.end(), prev + 1);
      if (it == p.end()) return m[word] = false;
      prev = *it;
    }

    return m[word] = true;
  }
};

File: leetcode-04/solutions/python3/0792.py
class Solution:
    def numMatchingSubseq(self, S: str, words: List[str]) -> int:
        def isMatch(word: str) -> int:
            if word in m:
                return m[word]

            prev = -1
            for c in word:
                index = bisect.bisect_left(dict[c], prev + 1)
                if index == len(dict[c]):
                    m[word] = 0
                    return 0
                prev = dict[c][index]
            m[word] = 1

            return 1

        m = {}
        dict = {chr(i + ord('a')): [] for i in range(26)}

        for i, c in enumerate(S):
            dict[c].append(i)

        return sum(map(isMatch, words))
